2023-02-26-binplz.md

```diff
@@ -0,0 +1,116 @@
+is statically linked.
+
+Behind the scenes, when you access a URL like `binplz.dev/tree`, the web server
+running on `binplz.dev` builds the derivation
+[`pkgsStatic.tree`](https://functor.tokyo/blog/2021-10-20-nix-cross-static) from
+Nixpkgs, pulls out the statically-linked `bin/tree` binary from the output, and
+returns it.
+
+`binplz.dev` should be able to build and serve any package from Nixpkgs.[^2]
+
+## Writing binplz
+
+Jonas and Hideaki joined me to build an initial version of binplz.  We created a
+[binplz repo](https://github.com/binplz/binplz.dev/) to hold our code,
+and Jonas quickly put together an initial implementation.
+
+The initial implementation did everything we required above, but our next task
+was getting it deployed.
+
+## Problems
+
+When trying to work out how to deploy this service, we ran into a few
+problems.
+
+When a user requests a binary like `tree`, binplz is going to have to build it
+at least once, since the Nixpkgs Hydra doesn't build and cache most things from
+`pkgsStatic`.  Most of our problems revolved around the following:
+
+1.  Where do we do the actually building?  A single build machine on AWS or
+    Hetzner might work, but it would be slow if a bunch of users request
+    derivations like Chromium and Firefox.  Auto-scaling might be better, but
+    it could get really expensive.[^1]
+
+2.  Where do we store the build outputs?  An S3 cache is an obvious choice, but
+    the store could get really expensive in certain cases.
+
+    For instance, if we allow users to build binaries from any arbitrary
+    Nixpkgs commit, our stores could blow up to tens or hundreds of terrabytes.
+    Or, what if there happens to be some sort of dynamic attr path in Nixpkgs a
+    user could use to keep building different derivations?
+
+3.  We probably need some sort of caching when builds fail, so that we don't
+    retry them. For instance, we don't want to repeatedly try to rebuild a huge
+    piece of software like PyTorch if we know it will always fail halfway
+    through.
+
+    This leads to a few interesting follow-up problems, like how do you tell
+    an actual build failure from a transient network problem, or hardware
+    problem?
+
+None of these problems are insurmountable, but we all sort of lost interest
+before we implemented solutions to each of them.
+
+If you're interested in more discussion on each of these, check out the
+[issue tracker](https://github.com/binplz/binplz.dev/issues).
+
+## Still a Good Idea
+
+I still think a service like binplz is a neat idea.  It would really show off
+the flexibility and power of Nix and Nixpkgs.  It would be a useful service if
+you need a certain binary in a pinch.  There are a bunch of different directions
+you could take it in to do other cool things.
+
+I hope someone decides to pick up this idea and drive it to completion!
+
+## Footnotes
+
+[^1]: We were seriously considering using [nixbuild.net](https://nixbuild.net/)
+    for this.
```

Comment:
I think you should say they were very willing to help out

Reply:
Good point.  I moved this from a footnote to an actual paragraph in the article in 7698efe.